### PR TITLE
Bug fixes for insurance card

### LINKED ISSUE #5953: old insurances showing "until Present" if latest insurance is Unassigned
### Describe the bug
From Dr Alexander Berlin, TX
Another important bug fix. In the new system, they got the insurance card wrong, with all insurances, including old insurances showing "until Present".

Here's the fix and it also introduces the concept of Self-Pay display for those times when a patient did not have insurance, both currently and in the past.

### To Reproduce

1. Add a '....'

### Client configuration

- Browser:
- OpenEMR version:
- Operating system:

### PR DESCRIPTION
**Pull request by juggernautsei**
Fixes #5953 

#### Short description of what this resolves:
From Dr Alexander Berlin, TX
Another important bug fix. In the new system, they got the insurance card wrong, with all insurances, including old insurances showing "until Present".

Here's the fix and it also introduces the concept of Self-Pay display for those times when a patient did not have insurance, both currently and in the past.

But wait there is more.
Unallocated amount. We have the ability to take a pre-payment on the front_payment screen, but it just doesn't show anywhere obvious, so that it could be applied to the account!

#### Changes proposed in this pull request:
Go to patient chart, put in a couple of insurances (primary and secondary) with an effective date of let’s say a year ago. Then add another primary insurance (as a separate test, set to Unassigned for primary) with an effective date of a month ago and set secondary to Unassigned for the same effective date a month ago.
Then refresh patient demographics by clicking on the patient name above. See what you see for insurances and their effective dates (insurance tab on top - Insurance from… until …).

### CODE REVIEW

File: library/patient.inc.php [hunk 1]
```diff
@@ -402,11 +402,49 @@ function getInsuranceDataByDate(
...
+        $total_amt = $row['pay_total'] - $row['global_amount'];
+        $rs = sqlQuery("SELECT sum(pay_amount) AS total_pay_amt FROM ar_activity WHERE session_id = ? AND pid = ? AND deleted IS NULL", array($row['session_id'], $pid));
+        $pay_amount = $rs['total_pay_amt'];
+        $unallocated += ($total_amt - $pay_amount);
+    }
+    return sprintf('%01.2f', $unallocated);
+}
+
+//ALB Added this function
+function getInsuranceNameByDate(
```
Comment (stephenwaite) [inline library/patient.inc.php:424]: not seeing this used anywhere

File: library/patient.inc.php [hunk 2]
```diff
@@ -630,7 +668,7 @@ function getPatientId($pid = "%", $given = "pid, id, lname, fname, mname, provid
 function getByPatientDemographics($searchTerm = "%", $given = "pid, id, lname, fname, mname, providerID, DATE_FORMAT(DOB,'%m/%d/%Y') as DOB_TS", $orderby = "lname ASC, fname ASC", $limit = "all", $start = "0")
 {
     $layoutCols = sqlStatement(
-        "SELECT field_id FROM layout_options WHERE form_id = 'DEM' AND field_id not like ? AND uor !=0",
+        "SELECT field_id FROM layout_options WHERE form_id = 'DEM' AND field_id not like ? AND uor ! = 0",
```
Comment (stephenwaite) [inline library/patient.inc.php:671]: can revert or could be 
```
"SELECT field_id FROM layout_options WHERE form_id = 'DEM' AND field_id not like ? AND uor != 0",
```

**Review by stephenwaite — COMMENTED** [review]
hi @juggernautsei , there's some code duplication that could be fixed in the else block of the self pay stuff and not sure if it'd be better to just add `ALB` as an author in the top docblock than to have ALB commented all over.

Haven't tested the unallocated patient balance.

File: library/patient.inc.php (hunk 1 above)
Comment (juggernautsei) [reply library/patient.inc.php:424]: The usage is coming in another commit. I have 4 more from him to post. 
I have done very little cleanup  on his work. I will update the top section and put him as an author. I was using the marking to make sure I got all the right stuff. I was going to wait until final draft and after testing before removing all the ALBs

File: library/patient.inc.php (hunk 2 above)
Comment (juggernautsei) [reply library/patient.inc.php]: @stephenwaite have you completed your testing?

Comment (stephenwaite) [reply library/patient.inc.php]: hi @juggernautsei, is it okay if I test after you push the other suff?

Comment (juggernautsei) [reply library/patient.inc.php]: Yes

Comment (stephenwaite) [reply library/patient.inc.php]: hi @juggernautsei, if you'd like I can push the changes here to try and help with that [community forum thread](https://community.open-emr.org/t/patient-insurance-history-build/17416/19?u=stephenwaite).

Comment (juggernautsei) [reply library/patient.inc.php]: yes, lets do that

**Comment by juggernautsei** [comment]
@stephenwaite where you going to add to this pull request?

**Comment by stephenwaite** [comment]
should be able to push something here soon, maybe as early as tomorrow

File: templates/patient/card/billing.html.twig [hunk 3]
```diff
@@ -2,9 +2,9 @@
 
 {% block content %}
 <div class="container">
-    <div class="row">
-        <div class="col-4">{{ "Patient Balance Due"|xlt }}</div>
-        <div class="col">{{ patientBalance|money|text }}</div>
+    <div class="row"> <!--ALB Changed this to red and bold -->
+        <div class="col-4 font-weight-bold"><font color='red'>{{ "Patient Balance Due"|xlt }}</font></div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:6]: Don't add `font-weight-bold` here - it's very overused throughout the UI and I am actively working on removing it. The `<font>` tag is long deprecated and there's no need to make this text red, leave it as it.

File: templates/patient/card/billing.html.twig [hunk 4]
```diff
@@ -2,9 +2,9 @@
...
+        <div class="col font-weight-bold"><font color='red'>{{ patientBalance|money|text }}</font></div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:7]: See above

File: templates/patient/card/billing.html.twig [hunk 5]
```diff
@@ -14,26 +14,34 @@
         <div class="col-4 font-weight-bold">{{ "Total Balance Due"|xlt }}</div>
         <div class="col font-weight-bold">{{ totalBalance|money|text }}</div>
     </div>
-    {% if billingNote %}
+
+    {% if unallocated > 0 %}{# ALB Show unallocated amount, if present #}
     <div class="row">
-        <div class="col-4">{{ "Billing Note"|xlt }}</div>
-        <div class="col">{{ billingNote|text }}</div>
+        <div class="col-4 font-weight-bold"><font color='#ee6600'>{{ "Unallocated Patient Pre-Payment"|xlt }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:20]: See above

File: templates/patient/card/billing.html.twig [hunk 6]
```diff
@@ -14,26 +14,34 @@
...
+        <div class="col font-weight-bold">{{ unallocated|money|text }}</font></div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:21]: See above

File: templates/patient/card/billing.html.twig [hunk 7]
```diff
@@ -14,26 +14,34 @@
...
+    </div>
+    {% endif %}
+
+    {% if billingNote %}
+    <div class="row"> <!--ALB Made all bold -->
+        <div class="col-4 font-weight-bold">{{ "Billing Note"|xlt }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:27]: See above

File: templates/patient/card/billing.html.twig [hunk 8]
```diff
@@ -14,26 +14,34 @@
...
+        <div class="col font-weight-bold">{{ billingNote|text }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:28]: See above

File: templates/patient/card/billing.html.twig [hunk 9]
```diff
@@ -14,26 +14,34 @@
...
     </div>
     {% endif %}
     {% if provider %}
     <div class="row">
-        <div class="col-4">{{ "Primary Insurance"|xlt }}</div>
-        <div class="col">{{ insName|text }}</div>
+        <div class="col-4 font-weight-bold">{{ "Primary Insurance"|xlt }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:33]: See above

File: templates/patient/card/billing.html.twig [hunk 10]
```diff
@@ -14,26 +14,34 @@
...
+        <div class="col font-weight-bold">{{ insName|text }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:34]: See above

File: templates/patient/card/billing.html.twig [hunk 11]
```diff
@@ -14,26 +14,34 @@
...
     </div>
         {% if copay %}
         <div class="row">
-            <div class="col-4">{{ "Copay"|xlt }}</div>
-            <div class="col">{{ copay|text }}</div>
+            <div class="col-4 font-weight-bold">{{ "Copay"|xlt }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:38]: See above

File: templates/patient/card/billing.html.twig [hunk 12]
```diff
@@ -14,26 +14,34 @@
...
+            <div class="col font-weight-bold">{{ copay|text }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:39]: See above

File: templates/patient/card/billing.html.twig [hunk 13]
```diff
@@ -14,26 +14,34 @@
...
         </div>
         {% endif %}
     <div class="row">
-        <div class="col-4">{{ "Effective Date"|xlt }}</div>
-        <div class="col">{{ effDate|shortDate|text }}</div>
+        <div class="col-4 font-weight-bold">{{ "Effective Date"|xlt }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:43]: See above

File: templates/patient/card/billing.html.twig [hunk 14]
```diff
@@ -14,26 +14,34 @@
...
+        <div class="col font-weight-bold">{{ effDate|shortDate|text }}</div>
```
Comment (robertdown) [inline templates/patient/card/billing.html.twig:44]: See above

**Review by robertdown — CHANGES_REQUESTED** [review]
Need to rollback a lot of the UI changes in billing twig. No `font-weight-bold` or use of the color red here. Multiple reasons for this. First, we use bold way too much throughout the UI, labels shouldn't be bolded, or, if they should be, they _all_ should be (and would be done in a theme, not hard-coded classes). Second, using red should indicate a **highly** critical piece of information. I don't think carrying a balance would fit that. Lastly, the overuse of bold and red colors lead to alarm fatigue throughout the user interface and any use needs to be heavily considered. An example of this is the MRD - very little color is used... except if a patient has a severe allergy, then we call a lot of attention to it. Reserving bold and red items for critical communication to the user helps prevent desensitizing the end user.

This is a soap box for me.... see #4739 :)

File: templates/patient/card/billing.html.twig (hunk 14 above)
Comment (juggernautsei) [reply templates/patient/card/billing.html.twig:44]: The doctor that did these changes is showing the age of his coding skills. I have shared with him that there are a lot of things he do that are out of date. If he is going to continue to code. He needs to update his skills or risk writing code that no longer works.

**Comment by juggernautsei** [comment]
Bump here also. Don't want this to get lost to the crowd. @robertdown

**Comment by robertdown** [comment]
Will look at this tonight. Been out sick for a few days

**Comment by juggernautsei** [comment]
@robertdown Thanks, I hope your recovery goes well.

**Comment by robertdown** [comment]
@juggernautsei I pushed a fix to the unallocated function.  Initially, you set `$unallocated = ''`' which was causing line 417 to throw a fatal error as it was trying to add a float to a string. The commit sets the initial `$unallocated = 0.0` which is a true float so further adding can move on without issue.

**Comment by robertdown** [comment]
@juggernautsei if you can resolve the conflicts I think we can bring this in

**Comment by juggernautsei** [comment]
I don't have the ability or access to resolve conflicts.

**Comment by robertdown** [comment]
Conflicts are resolved on the incoming PR branch
* checkout master
* pull updates from upstream
* merge master into insurance_card_information_adjustments branch
* merge won't be automatic, you'll resolve the discrepancies there
* commit the newly updated file
* push

**Comment by bradymiller** [comment]
@juggernautsei , This doesn't look right after the merge. Looks like reverted wrong changes (kept master rather than your own). Let me know if you want me to bring in another PR with your code that is ready for codebase.

**Comment by bradymiller** [comment]
This feature was brought in via subsequent pull request: https://github.com/openemr/openemr/pull/6105

**Pull request closed without merging**